### PR TITLE
Implement modules for business-logic rate limiting

### PR DESCRIPTION
**Pull request by tgregory-block**
### Motiviation
To give services an easy to implement alternative to `cash-permit` for business logic rate limiting, backed by infra dedicated to that service. 

This PR implements token bucket based rate limiting locally in `wisp` (and using that interface, in `misk`), using a library called [bucket4j](https://github.com/bucket4j/bucket4j). There are a few layers that are built up, until we can install a module in a misk app and use it for rate limiting. 
* First we stand up a `RateLimiter` interface for rate limit token operations
* In this same commit, we add an implementation of `RateLimiter` backed by Bucket4j that takes a `ProxyManager`, which is Bucket4j's generic interface for backing stores such as redis and dynamo.
* Next we wire up two `misk` modules, each injecting an implementation of `RateLimiter`, one backed by Redis and one backed by DynamoDB
* Finally, we implement an example rate limited action in the sample misk app, testing both DynamoDB and Redis backends

### CODE REVIEW

File: misk-rate-limiting-bucket4j-dynamodb-v1/src/test/kotlin/misk/ratelimiting/bucket4j/dynamodb/v1/DynamoDBV1RateLimiterTest.kt [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+      )
+      install(DynamoDbV1Bucket4jRateLimiterModule("rate_limit_buckets"))
+      install(MiskTestingServiceModule())
+      install(DeploymentModule(TESTING))
+    }
+
+    @Provides @Singleton
+    // In prod this is provided by Skim
+    fun provideMeterRegistry(collectorRegistry: CollectorRegistry): MeterRegistry {
+      return PrometheusMeterRegistry(
```
Comment (ThePumpingLemma) [inline misk-rate-limiting-bucket4j-dynamodb-v1/src/test/kotlin/misk/ratelimiting/bucket4j/dynamodb/v1/DynamoDBV1RateLimiterTest.kt:45]: Would recommend using `SimpleMeterRegistry` if you don't need to test anything prometheus specific.

File: misk-rate-limiting-bucket4j-dynamodb-v1/build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+dependencies {
+  api(project(":misk-aws-dynamodb"))
+  api(project(":misk-inject"))
+  api(project(":wisp:wisp-rate-limiting:bucket4j"))
+
+  implementation(project(":misk-inject"))
+  implementation(Dependencies.bucket4jDynamoDbV1)
+
+  testImplementation(Dependencies.micrometerPrometheus)
```
Comment (ThePumpingLemma) [inline misk-rate-limiting-bucket4j-dynamodb-v1/build.gradle.kts:20]: If you use `SimpleMeterRegistry`, you can depend on `micrometer-core` instead.

File: misk-rate-limiting-bucket4j-dynamodb-v1/src/test/kotlin/misk/ratelimiting/bucket4j/dynamodb/v1/DynamoDBV1RateLimiterTest.kt (hunk 1 above)
Comment (tgregory-block) [reply misk-rate-limiting-bucket4j-dynamodb-v1/src/test/kotlin/misk/ratelimiting/bucket4j/dynamodb/v1/DynamoDBV1RateLimiterTest.kt]: Thanks, fixed this everywhere I added it 🙏

File: buildSrc/src/main/kotlin/Dependencies.kt [hunk 3]
```diff
@@ -17,6 +17,10 @@ object Dependencies {
   val awsSqs = "com.amazonaws:aws-java-sdk-sqs:1.12.520"
   val bouncycastle = "org.bouncycastle:bcprov-jdk15on:1.70"
   val bouncycastlePgp = "org.bouncycastle:bcpg-jdk15on:1.70"
+  val bucket4jCore = "com.bucket4j:bucket4j-core:8.5.0"
```
Comment (ThePumpingLemma) [inline buildSrc/src/main/kotlin/Dependencies.kt:20]: FYI, `8.6.0` just came out.

File: wisp/wisp-rate-limiting/src/main/kotlin/wisp/ratelimiting/RateLimitConfiguration.kt [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  /**
+   * The maximum number of tokens that can accumulate in the bucket
+   */
+  val capacity: Long
+
+  /**
+   * A name identifying the configuration, e.g. "OriginationFileProcessing" for a rate limit
+   * configuration governing the rate at which origination files can be processed
+   */
+  val name: String
```
Comment (ThePumpingLemma) [inline wisp/wisp-rate-limiting/src/main/kotlin/wisp/ratelimiting/RateLimitConfiguration.kt:18]: nit: should consistently have (or omit) whitespace between each field?

File: wisp/wisp-rate-limiting/src/main/kotlin/wisp/ratelimiting/RateLimiter.kt [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+ *
+ * * `token` - a unit of rate limit capacity. When we go to perform a rate limited operation, we
+ * attempt to consume a token from the bucket. If successful, we can perform the operation.
+ *
+ * See also: https://en.wikipedia.org/wiki/Token_bucket
+ */
+interface RateLimiter {
+  /**
+   * Consumes [amount] tokens from the bucket associated with the given key
+   * TODO Confirm the exception type is correct
```
Comment (ThePumpingLemma) [inline wisp/wisp-rate-limiting/src/main/kotlin/wisp/ratelimiting/RateLimiter.kt:22]: Do we still need to confirm this?

**Review by ThePumpingLemma — APPROVED** [review]

File: buildSrc/src/main/kotlin/Dependencies.kt (hunk 3 above)
Comment (tgregory-block) [reply buildSrc/src/main/kotlin/Dependencies.kt:20]: Indeed, it only had a fix in functionality this doesn't use (local in-memory buckets). I'm [planning on doing a fast follow](https://cash.slack.com/archives/C01M5M297T7/p1699292746150629?thread_ts=1698944488.920549&cid=C01M5M297T7) to `8.7.0` to get `UnifiedJedis` support, for better compatibility with #2997

File: wisp/wisp-rate-limiting/src/main/kotlin/wisp/ratelimiting/RateLimiter.kt (hunk 5 above)
Comment (tgregory-block) [reply wisp/wisp-rate-limiting/src/main/kotlin/wisp/ratelimiting/RateLimiter.kt:22]: This is actually incorrect, it surfaces whatever exception the underlying library throws (e.g. `JedisException` subclasses). Will add a note about that in the doc string

**Pull request merged**